02-peer-protocol.md

```diff
@@ -49,11 +68,515 @@ transaction is confirmed are also not persistent - until you know the script
+    - the `serial_id` is already included in the transaction
+    - the `serial_id` has the wrong parity
+    - if has received 4096 `tx_add_input` messages during this negotiation
+
+#### Rationale
+
+Each node must know the set of the transaction inputs. The *non-initiator*
+MAY omit this message.
+
+`serial_id` is a randomly chosen number which uniquely identifies this input.
+Inputs in the constructed transaction MUST be sorted by `serial_id`.
+
+`prevtx` is the serialized transaction that contains the output
+this input spends. Used to verify that the input is non-malleable.
+
+`prevtx_vout` is the index of the output being spent.
+
+`sequence` is the sequence number of this input: it must signal
+replaceability, and the same value should be used across implementations
+to avoid on-chain fingerprinting.
+
+#### Liquidity griefing
+
+When sending `tx_add_input`, senders have no guarantee that the remote node
+will complete the protocol in a timely manner. Malicious remote nodes could
+delay messages or stop responding, which can result in a partially created
+transaction that cannot be broadcast by the honest node. If the honest node
+is locking the corresponding UTXO exclusively for this remote node, this can
+be exploited to lock up the honest node's liquidity.
+
+It is thus recommended that implementations keep UTXOs unlocked and actively
+reuse them in concurrent sessions, which guarantees that transactions created
+with honest nodes double-spend pending transactions with malicious nodes at
+no additional cost for the honest node.
+
+Unfortunately, this will also create conflicts between concurrent sessions
+with honest nodes. This is a reasonable trade-off though because:
+
+* on-chain funding attempts are relatively infrequent operations
+* honest nodes should complete the protocol quickly, reducing the risk of
+  conflicts
+* failed attempts can simply be retried at no cost
+
+### The `tx_add_output` Message
+
+This message adds a transaction output.
+
+1. type: 67 (`tx_add_output`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`u64`:`serial_id`]
+    * [`u64`:`sats`]
+    * [`u16`:`scriptlen`]
+    * [`scriptlen*byte`:`script`]
+
+#### Requirements
+
+Either node:
+  - MAY omit this message
+
+The sending node:
+  - MUST add all sent outputs to the transaction
+  - if is the *initiator*:
+    - MUST send even `serial_id`s
+  - if is the *non-initiator*:
+    - MUST send odd `serial_id`s
+
+The receiving node:
+  - MUST add all received outputs to the transaction
+  - MUST accept P2WSH, P2WPKH, P2TR `script`s
+  - MAY fail the negotiation if `script` is non-standard
+  - MUST fail the negotiation if:
+    - the `serial_id` is already included in the transaction
+    - the `serial_id` has the wrong parity
+    - it has received 4096 `tx_add_output` messages during this negotiation
+    - the `sats` amount is less than the `dust_limit`
+    - the `sats` amount is greater than 2,100,000,000,000,000 (`MAX_MONEY`)
```

Comment:
we should probably not only check that the `sats` amount of this output is less than `MAX_MONEY`, but the sum of all the outputs thus far?

alternatively, we could make the summation check prior to sending `tx_signatures`?

Reply:
we generally defer full checks until the `tx_complete` step (eg a peer could have an invalid tx inflight at any point prior to sending tx_complete.) these checks happens prior to sending commitment signatures (which is before `tx-sigs`).

on an output by output basis, it's reasonable to bound to "not an outlandish amount of bitcoin". at the `tx_complete` finalization stage, we will also check that they've provided enough inputs to cover whatever amount of outputs they're claiming they can cover.